### PR TITLE
ci: sgx ci/cd

### PR DESCRIPTION
**Pull request by maceip**
ci: tee ci/cd to build and deploy sgx notary-server
Note: the cd script is temporary and should not live beyond 2024. I will replace the janky reverse proxy update.sh and docker kill / start  logic  with k8s / compose / helm

### CODE REVIEW

File: cd-scripts/tee/azure/Caddyfile [hunk 1]
```diff
@@ -0,0 +1,51 @@
+  key_type p256
```
Comment (heeckhau) [inline cd-scripts/tee/azure/Caddyfile:1]: It looks like there is a line missing here?

File: .github/workflows/tee-cd.yml [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+          run: |
+            echo "SCCACHE_GHA_ENABLED=true" >> $GITHUB_ENV
+            echo "RUSTC_WRAPPER=sccache" >> $GITHUB_ENV
+        - name: reverse proxy port
+          run: echo "${{needs.update-reverse-proxy.outputs.teeport}}" | tee >> $GITHUB_STEP_SUMMARY
+
+        - name: get measurement
+          working-directory: ${{ github.workspace }}/crates/notary/server/fixture/tee
+          run: |
+             curl https://sh.rustup.rs -sSf | sh -s -- -y
```
Comment (heeckhau) [inline .github/workflows/tee-cd.yml:101]: Move to script

File: .github/workflows/tee-cd.yml [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        - name: auth to registry
+          uses: docker/login-action@v3
+          with:
+            registry: notaryserverbuilds.azurecr.io
+            username: notaryserverbuilds
+            password: ${{ secrets.AZURE_CR_BUILDS_PW }}
+        - name: get code
+          uses: actions/checkout@v4
+          with:
+            repository: tlsnotary/tlsn
```
Comment (heeckhau) [inline .github/workflows/tee-cd.yml:146]: remove?

File: .github/workflows/tee-cd.yml [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        -
+          name: Set up Docker Buildx
+          uses: docker/setup-buildx-action@v3
+        - name: Build and push
+         
+   
+          uses: docker/build-push-action@v6
+          with:
+            context: ${{ github.workspace }}/crates/notary/server/fixture/tee
+            # testing:
```
Comment (heeckhau) [inline .github/workflows/tee-cd.yml:167]: ???

File: .github/workflows/tee-cd.yml [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+          uses: mozilla-actions/sccache-action@v0.0.6
+        - name: set rust env for scc
+          if: github.event_name != 'release'
+          # && github.event_name != 'workflow_dispatch'
+          run: |
+            echo "SCCACHE_GHA_ENABLED=true" >> $GITHUB_ENV
+            echo "RUSTC_WRAPPER=sccache" >> $GITHUB_ENV
+        -
+          name: Set up Docker Buildx
+          uses: docker/setup-buildx-action@v3
```
Comment (heeckhau) [inline .github/workflows/tee-cd.yml:160]: No need to add directory of dockerfile here? context argument?

**Review by heeckhau — CHANGES_REQUESTED** [review]

**Review by yuroitaki — COMMENTED** [review]
🚀 great stuff especially with the caddy reverse proxy! pretty much covered all the main features the current aws deployment setup has; the only missing piece is stopping the oldest version in the server upon a new release — currently we only maintain the 3 latest versions

File: cd-scripts/tee/azure/updateproxy.sh [hunk 6]
```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
+
+# Variables (Update these as needed)x
+CADDYFILE=${1:-/etc/caddy/Caddyfile}  # Path to your Caddyfile
+GIT_COMMIT_HASH=${2:-dev}
+BASE_PORT=6061                      # The starting port for your reverse_proxy directives
```
[the remark this answers is not in the document]
Comment (maceip) [reply cd-scripts/tee/azure/updateproxy.sh:6]: - how many releases would we be able to serve before we hit a well-known-service port?
- why would there be a conflict?

File: .github/workflows/tee-cd.yml [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+#on:
+#  release:
+#    types: [published]
+#    branches:
+#      - 'releases/**'
+
+env:
+  GIT_COMMIT_HASH: ${{ github.event.pull_request.head.sha || github.sha }}
+  GIT_COMMIT_TIMESTAMP: ${{ github.event.repository.updated_at}}
```
[the remark this answers is not in the document]
Comment (maceip) [reply .github/workflows/tee-cd.yml:20]: - help me understand how you are using the timestamp value downstream
- how would this workflow get triggered in a way where the timestamp value was wrong?   
     - is someone updating the description not a relevant timestamp update?

File: .github/workflows/tee-ci.yml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        credentials:
+          username: notaryserverbuilds
+          password: ${{ secrets.AZURE_CR_BUILDS_PW }}
+        env:
+          GIT_COMMIT_HASH: ${{ github.event.pull_request.head.sha || github.sha }}
+      steps:
+        - name: get code
+          uses: actions/checkout@v4
+        - name: sccache
+          if: github.event_name != 'release'
```
[the remark this answers is not in the document]
Comment (maceip) [reply .github/workflows/tee-ci.yml:28]: the spirit of this line is to make sure caching is off if the event that triggered the workflow is 'release'. are you suggesting that this does not work?

File: .github/workflows/tee-cd.yml [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+        id: portbump
+        env:
+          RELEASE_TAG: ${{ github.event.release.tag_name ||  inputs.ref }}
+
+        run: |
+          NEXT_PORT=$(bash cd-scripts/tee/azure/updateproxy.sh 'cd-scripts/tee/azure/Caddyfile' $RELEASE_TAG)
+          echo "newport=$NEXT_PORT" >> $GITHUB_OUTPUT
+          
+      - name: Deploy updated Caddyfile to server
+        if: ${{ steps.portbump.outputs.deploy == 'new' }}
```
[the remark this answers is not in the document]
Comment (maceip) [reply .github/workflows/tee-cd.yml:42]: - great catch!

pushed my local commit that has it, here => https://github.com/tlsnotary/tlsn/commit/90c495b65a7dfcb261d858535eae806245836fc9#diff-1754ed6057c6ea37a996f019ee28c3b3267a41031f91187330ff6e19747618e7R80

File: cd-scripts/tee/azure/updateproxy.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (yuroitaki) [reply cd-scripts/tee/azure/updateproxy.sh:6]: Ok a quick check online gives me this list for port from 6061 to 6111 haha
image

None of them seem to be well-known to me? I guess we can also just consciously not use any port in the nearby range from 6061 in the server, so ya should be fine 👍

File: .github/workflows/tee-cd.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (yuroitaki) [reply .github/workflows/tee-cd.yml:20]: The timestamp ll be returned as part of the `/info` endpoint:
```
{
    "version": "0.1.0-alpha.7",
    "publicKey": "-----BEGIN PUBLIC KEY----- ...",
    "gitCommitHash": "99e02fb388c2e791d03c3426bd0411395bfaf453",
    "gitCommitTimestamp": "2024-10-14T20:23:19+02:00"
}
```
It's just an extra metadata to let end user know 'when' the code running in the server was last updated, as it's the timestamp of "gitCommitHash", the head commit.

So ya it shouldn't be the timestamp when someone updates the repo description / create a wiki page some time after the head commit is committed.

**Comment by yuroitaki** [comment]
For the other comments on the release trigger: instead of the release-triggered ci/cd approach, tlsn uses the tag-triggered approach — I think we prefer to ensure the new version is successfully deployed and running before the release is announced.

Anyway, as discussed yesterday in the meeting, since this tee-ci/cd will one day replace [`cd-server.yml`](https://github.com/tlsnotary/tlsn/blob/main/.github/workflows/cd-server.yml) (the current aws deployment), I think this ci/cd can just run upon successful completion of the [`ci.yml`](https://github.com/tlsnotary/tlsn/blob/main/.github/workflows/ci.yml) integration test job? like [this](https://github.com/tlsnotary/tlsn/blob/main/.github/workflows/cd-server.yml#L47-L56)

File: .github/workflows/tee-cd.yml [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+#    types: [published]
+#    branches:
+#      - 'releases/**'
+
+env:
+  GIT_COMMIT_HASH: ${{ github.event.pull_request.head.sha || github.sha }}
+  GIT_COMMIT_TIMESTAMP: ${{ github.event.repository.updated_at}}
+
+jobs:
+  update-reverse-proxy:
```
[the remark this answers is not in the document]
Comment (maceip) [reply .github/workflows/tee-cd.yml:23]: - going to refactor this to be a core part of cd, needs more testing 

- initially this was built as a sidecar deployment, hence the trigger being post-release.

**Comment by maceip** [comment]
> For the other comments on the release trigger: instead of the release-triggered ci/cd approach, tlsn uses the tag-triggered approach — I think we prefer to ensure the new version is successfully deployed and running before the release is announced.
> 
> Anyway, as discussed yesterday in the meeting, since this tee-ci/cd will one day replace [`cd-server.yml`](https://github.com/tlsnotary/tlsn/blob/main/.github/workflows/cd-server.yml) (the current aws deployment), I think this ci/cd can just run upon successful completion of the [`ci.yml`](https://github.com/tlsnotary/tlsn/blob/main/.github/workflows/ci.yml) integration test job? like [this](https://github.com/tlsnotary/tlsn/blob/main/.github/workflows/cd-server.yml#L47-L56)

100

i just commented somewhere above im refactoring it to fit in with cd.yml, and not as a sidecar/lesser cd -- needs more testing but will have a cut tmmrw and will ping you

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -138,4 +138,11 @@ jobs:
...
           files: lcov.info
-          fail_ci_if_error: true
\ No newline at end of file
+          fail_ci_if_error: true
+  trigger-deployment:
+    # doing this here due to feedback @ https://github.com/tlsnotary/tlsn/pull/631#issuecomment-2415806267
+    needs: tests-integration
+    uses: ./.github/workflows/tee-cd.yml
+    with:
+      # what this is supposed to do -> $ref is the tag: e.g., v0.1.0-alpha.7; pass the $ref string to the cd script and update reverse proxy / deploy
```
Comment (heeckhau) [inline .github/workflows/ci.yml:147]:
```suggestion
      # pass the $ref string to the cd script and update reverse proxy / deploy
      # $ref is the tag: e.g., v0.1.0-alpha.7
```

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -138,4 +138,11 @@ jobs:
         with:
           token: ${{ secrets.CODECOV_TOKEN }}
           files: lcov.info
-          fail_ci_if_error: true
\ No newline at end of file
+          fail_ci_if_error: true
+  trigger-deployment:
```
Comment (heeckhau) [inline .github/workflows/ci.yml:142]:
```suggestion
  trigger-tee-deployment:
```

File: .github/workflows/tee-cd.yml [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+        description: 'git branch'
+        required: false
+        default: 'dev'
+        type: string
+
+#on:
+#  release:
+#    types: [published]
+#    branches:
+#      - 'releases/**'
```
Comment (heeckhau) [inline .github/workflows/tee-cd.yml:16]: remove or activate?

**Review by heeckhau — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 12 above)
Comment (yuroitaki) [inline .github/workflows/ci.yml:142]: wouldn't this deployment run everytime a PR is raised? as the trigger condition in this `ci.yml` is
```
on:
  push:
    branches:
      - dev
    tags:
      - "[v]?[0-9]+.[0-9]+.[0-9]+*"
  pull_request:

```

Comment (heeckhau) [reply .github/workflows/ci.yml:142]: We can add this to the `trigger-tee-deployment` job:
```
    # only for dev branch and tags
    if: github.event_name == 'push' && (startsWith(github.ref, 'refs/tags/') || github.ref == 'refs/heads/dev')
```

**Pull request merged**